Rubberduck.Refactorings/Common/IModuleRewriterExtensions.cs

```diff
@@ -0,0 +1,101 @@
+﻿using Antlr4.Runtime;
+using Rubberduck.Parsing;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Rewriter;
+using Rubberduck.Parsing.Symbols;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Rubberduck.Refactorings.Common
+{
+    public static class IModuleRewriterExtensions
+    {
+        /// <summary>
+        /// Removes variable declaration and subsequent <c>VBAParser.EndOfStatementContext</c>
+        /// depending on the <paramref name="removeEndOfStmtContext"/> flag.
+        /// This function is intended to be called only once per rewriter within a given <c>ModuleRewriteSession</c>.  
+        /// </summary>
+        /// <remarks>
+        /// Calling this function with <paramref name="removeEndOfStmtContext"/> defaulted to <c>true</c>
+        /// avoids leaving residual newlines between the deleted declaration and the next declaration. 
+        /// The one-time call constraint is required for scenarios where variables to delete are declared in a list.  Specifically,
+        /// the use case where all the variables in the list are to be removed.
+        /// If the variables to remove are not declared in a list, then this function can be called multiple times.
+        /// </remarks>
+        public static void RemoveVariables(this IModuleRewriter rewriter, IEnumerable<VariableDeclaration> toRemove, bool removeEndOfStmtContext = true)
+        {
+            if (!toRemove.Any()) { return; }
+
+            var fieldsByListContext = toRemove.Distinct()
+                                                .GroupBy(f => f.Context.GetAncestor<VBAParser.VariableListStmtContext>());
+
+            foreach (var fieldsGroup in fieldsByListContext)
+            {
+                var variables = fieldsGroup.Key.children.Where(ch => ch is VBAParser.VariableSubStmtContext);
+                if (variables.Count() == fieldsGroup.Count())
+                {
+                    if (fieldsGroup.First().ParentDeclaration.DeclarationType.HasFlag(DeclarationType.Module))
```

Comment:
This is a second iteration of `fieldsGroup`. Accordingly, it should be materialized via `ToList` where it is defined.

Reply:
I understand the comment..though I am curious as how to best materialize the `List` since `fieldsGroup` is defined within a `foreach (var fieldsGroup in fieldsByListContext)` statement.  Currently, I've added `var fieldsToDelete = fieldsGroup.ToList();`  and use `fieldsToDelete` for the `Count()` and `First()` calls which (I believe) resolves the comment.

Is there another more preferred way in this scenario?...perhaps using `ToLookup`  to instantiate `fieldsByListContext` instead of `GroupBy`  resolves this as well without calling `ToList` on `fieldsGroup`?

Reply:
That resolves the issue. You could alternatively use `ToDictionary` on `fieldsByListContext`. There should be an extension method turning it into a `Dictionary<TKey, List<TValue>>`.